middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -460,123 +368,158 @@ and let_cont_inline env res k h body =
+       via the raising of exceptions, whereas other continuations are reached
+       using a normal jump. *)
+    if is_exn_handler
+    then let_cont_exn_handler env res k body vars handler ~catch_id arity
+    else
+      let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
+      let body, res = expr env res body in
+      ( C.create_ccatch ~rec_flag:false ~body
+          ~handlers:[C.handler ~dbg catch_id vars handler],
+        res )
+  in
+  wrap cmm, res
+
+(* Exception continuations are translated using delayed Ctrywith blocks. The
+   exception handler parts of these blocks are identified by the [catch_id]s.
+
+   Additionally, exception continuations can have extra args, which are passed
+   through the try-with using mutable Cmm variables. Thus the exception handler
+   must first read the contents of those extra args (eagerly, in order to
+   minmize the lifetime of the mutable variables). *)
+and let_cont_exn_handler env res k body vars handler ~catch_id arity =
+  let exn_var, extra_params =
+    match vars with
+    | (v, _) :: rest -> v, rest
+    | [] ->
+      (* See comment on [translate_raise], above. *)
+      Misc.fatal_errorf
+        "Exception continuation %a should have at least one argument"
+        Continuation.print k
+  in
+  let env_body, mut_vars = Env.add_exn_handler env k arity in
   let handler =
-    (* wrap the exn handler with reads of the mutable variables *)
+    (* Wrap the exn handler with reads of the mutable variables *)
     List.fold_left2
-      (fun handler (v, _) (p, _) ->
-        C.letin p ~defining_expr:(C.var v) ~body:handler)
-      handle extra_vars extra_params
+      (fun handler (mut_var, _) (extra_param, _) ->
+        C.letin extra_param ~defining_expr:(C.var mut_var) ~body:handler)
+      handler mut_vars extra_params
   in
   let body, res = expr env_body res body in
   let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
-  let trywith = C.trywith ~dbg ~kind:(Delayed id) ~body ~exn_var ~handler () in
-  (* define and initialize the mutable cmm variables used by an exn extra
-     args *)
+  let trywith =
+    C.trywith ~dbg ~kind:(Delayed catch_id) ~body ~exn_var ~handler ()
+  in
+  (* Define and initialize the mutable Cmm variables for extra args *)
   let cmm =
     List.fold_left
-      (fun cmm (v, k) ->
+      (fun cmm (mut_var, (kind : K.t)) ->
         (* CR mshinwell: Fix [provenance] *)
-        let v = Backend_var.With_provenance.create ?provenance:None v in
-        C.letin_mut v (C.machtype_of_kind k) (default_of_kind ~dbg k) cmm)
-      trywith extra_vars
+        let mut_var =
+          Backend_var.With_provenance.create ?provenance:None mut_var
+        in
+        let dummy_value =
+          match kind with
+          | Value -> C.int ~dbg 1
+          | Naked_number Naked_float -> C.float ~dbg 0.
+          | Naked_number Naked_int64 when Target_system.is_32_bit ->
```

Comment:
now that #696 has been merged (or soon to be), this conditional can probably be removed

Reply:
Fixed